File: serialization/java/text/lib/src/main/java/com/microsoft/kiota/serialization/TextSerializationWriter.java
package com.microsoft.kiota.serialization;

import com.microsoft.kiota.serialization.SerializationWriter;
import com.microsoft.kiota.serialization.ValuedEnum;
import com.microsoft.kiota.serialization.Parsable;

import java.lang.Enum;
import java.lang.reflect.Field;
import java.lang.UnsupportedOperationException;
import java.math.BigDecimal;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStreamWriter;
import java.time.LocalDate;
import java.time.LocalTime;
import java.time.OffsetDateTime;
import java.time.Period;
import java.time.format.DateTimeFormatter;
import java.util.Base64;
import java.util.EnumSet;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;
import java.util.function.Consumer;
import java.util.function.BiConsumer;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

public class TextSerializationWriter implements SerializationWriter {
    private final static String NoStructuredDataMessage = "text does not support structured data";
    private final ByteArrayOutputStream stream = new ByteArrayOutputStream();
    private final OutputStreamWriter writer;
    private boolean written;
    public TextSerializationWriter() {
        this.writer = new OutputStreamWriter(this.stream);
    }
    public void writeStringValue(final String key, final String value) {
        if(key != null && !key.isEmpty())
            throw new UnsupportedOperationException(NoStructuredDataMessage);
        if(value != null && !value.isEmpty())
            if(written)
                throw new UnsupportedOperationException("a value was already written for this serialization writer, text content only supports a single value");
            else {
                written = true;
                try {
                    writer.write(value);
                } catch (IOException e) {
                    throw new RuntimeException(e);
                }
            }
    }
    public void writeBooleanValue(final String key, final Boolean value) {
        if(value != null)
            writeStringValue(key, value.toString());
    }
    public void writeShortValue(final String key, final Short value) {
        if(value != null)
            writeStringValue(key, value.toString());
    }
    public void writeByteValue(final String key, final Byte value) {
        if(value != null)
            writeStringValue(key, value.toString());
    }
    public void writeBigDecimalValue(final String key, final BigDecimal value) {
        if(value != null)
            writeStringValue(key, value.toString());
    }
    public void writeIntegerValue(final String key, final Integer value) {
        if(value != null)
            writeStringValue(key, value.toString());
    }
    public void writeFloatValue(final String key, final Float value) {
        if(value != null)
            writeStringValue(key, value.toString());
    }
    public void writeDoubleValue(final String key, final Double value) {
        if(value != null)
            writeStringValue(key, value.toString());
    }
    public void writeLongValue(final String key, final Long value) {
        if(value != null)
            writeStringValue(key, value.toString());
    }
    public void writeUUIDValue(final String key, final UUID value) {
        if(value != null)
            writeStringValue(key, value.toString());
    }
    public void writeOffsetDateTimeValue(final String key, final OffsetDateTime value) {
        if(value != null)
            writeStringValue(key, value.format(DateTimeFormatter.ISO_OFFSET_DATE_TIME));
    }
    public void writeLocalDateValue(final String key, final LocalDate value) {
        if(value != null)
            writeStringValue(key, value.format(DateTimeFormatter.ISO_LOCAL_DATE));
    }
    public void writeLocalTimeValue(final String key, final LocalTime value) {
        if(value != null)
            writeStringValue(key, value.format(DateTimeFormatter.ISO_LOCAL_TIME));
    }
    public void writePeriodValue(final String key, final Period value) {
        if(value != null)
            writeStringValue(key, value.toString());
    }
    public <T> void writeCollectionOfPrimitiveValues(final String key, final Iterable<T> values) {
        throw new UnsupportedOperationException(NoStructuredDataMessage);
    }
    public <T extends Parsable> void writeCollectionOfObjectValues(final String key, final Iterable<T> values) {
        throw new UnsupportedOperationException(NoStructuredDataMessage);
    }
    public <T extends Enum<T>> void writeCollectionOfEnumValues(@Nullable final String key, @Nullable final Iterable<T> values) {
        throw new UnsupportedOperationException(NoStructuredDataMessage);
    }
    public <T extends Parsable> void writeObjectValue(final String key, final T value) {
        throw new UnsupportedOperationException(NoStructuredDataMessage);
    }
    public <T extends Enum<T>> void writeEnumSetValue(@Nullable final String key, @Nullable final EnumSet<T> values) {
        throw new UnsupportedOperationException(NoStructuredDataMessage);
    }
    public <T extends Enum<T>> void writeEnumValue(@Nullable final String key, @Nullable final T value) {
        if(value != null) {
            writeStringValue(key, getStringValueFromValuedEnum(value));
        }
    }
    public void writeNullValue(@Nullable final String key) {
        writeStringValue(null, "null");
    }
    private <T extends Enum<T>> String getStringValueFromValuedEnum(final T value) {
        if(value instanceof ValuedEnum) {
            final ValuedEnum valued = (ValuedEnum)value;
            return valued.getValue();
        } else return null;
    }
    public InputStream getSerializedContent() {
        try {
            this.writer.flush();
            return new ByteArrayInputStream(this.stream.toByteArray());
            //This copies the whole array in memory could result in memory pressure for large objects, we might want to replace by some kind of piping in the future
        } catch (IOException ex) {
            throw new RuntimeException(ex);
        }
    }
    public void close() throws IOException {
        this.writer.close();
        this.stream.close();
    }
    public void writeAdditionalData(@Nonnull final Map<String, Object> value) {
        throw new UnsupportedOperationException(NoStructuredDataMessage);
    }
    public Consumer<Parsable> getOnBeforeObjectSerialization() {
        return this.onBeforeObjectSerialization;
    }
    public Consumer<Parsable> getOnAfterObjectSerialization() {
        return this.onAfterObjectSerialization;
    }
    public BiConsumer<Parsable, SerializationWriter> getOnStartObjectSerialization() {
        return this.onStartObjectSerialization;
    }
    private Consumer<Parsable> onBeforeObjectSerialization;
    public void setOnBeforeObjectSerialization(final Consumer<Parsable> value) {
        this.onBeforeObjectSerialization = value;
    }
    private Consumer<Parsable> onAfterObjectSerialization;
    public void setOnAfterObjectSerialization(final Consumer<Parsable> value) {
        this.onAfterObjectSerialization = value;
    }
    private BiConsumer<Parsable, SerializationWriter> onStartObjectSerialization;
    public void setOnStartObjectSerialization(final BiConsumer<Parsable, SerializationWriter> value) {
        this.onStartObjectSerialization = value;
    }
    public void writeByteArrayValue(@Nullable final String key, @Nonnull final byte[] value) {
        if(value != null)
            this.writeStringValue(key, Base64.getEncoder().encodeToString(value));
    }
}

File: serialization/php/text/src/TextSerializationWriter.php
<?php
/**
 * Copyright (c) Microsoft Corporation.  All Rights Reserved.
 * Licensed under the MIT License.  See License in the project root
 * for license information.
 */


namespace Microsoft\Kiota\Serialization\Text;


use DateInterval;
use DateTime;
use GuzzleHttp\Psr7\Utils;
use Microsoft\Kiota\Abstractions\Enum;
use Microsoft\Kiota\Abstractions\Serialization\Parsable;
use Microsoft\Kiota\Abstractions\Serialization\SerializationWriter;
use Microsoft\Kiota\Abstractions\Types\Byte;
use Microsoft\Kiota\Abstractions\Types\Date;
use Microsoft\Kiota\Abstractions\Types\Time;
use Psr\Http\Message\StreamInterface;

/**
 * Class TextSerializationWriter
 * @package Microsoft\Kiota\Serialization\Text
 * @copyright 2022 Microsoft Corporation
 * @license https://opensource.org/licenses/MIT MIT License
 */
class TextSerializationWriter implements SerializationWriter
{
    /**
     * @var StreamInterface
     */
    private StreamInterface $stream;

    /**
     * @var callable|null
     */
    private $onBeforeObjectSerialization = null;
    /**
     * @var callable|null
     */
    private $onAfterObjectSerialization = null;
    /**
     * @var callable|null
     */
    private $onStartObjectSerialization = null;

    /**
     * Create a TextSerializationWriter
     */
    public function __construct()
    {
        $this->stream = Utils::streamFor();
    }

    public function __destruct()
    {
        $this->stream->close();
    }

    /**
     * @inheritDoc
     */
    public function writeStringValue(?string $key, ?string $value): void
    {
        if ($key) {
            throw new \InvalidArgumentException('Keys not supported for text/plain content type');
        }
        if ($this->stream->getSize()) {
            throw new \RuntimeException('A value was already written for this serialization writer. Text content only allows a single value');
        }
        (!$value) ? $this->stream->write('') : $this->stream->write($value);
    }

    /**
     * @inheritDoc
     */
    public function writeBooleanValue(?string $key, ?bool $value): void
    {
        ($value) ? $this->writeStringValue($key, 'true') : $this->writeStringValue($key, 'false');
    }

    /**
     * @inheritDoc
     */
    public function writeFloatValue(?string $key, ?float $value): void
    {
        $this->writeStringValue($key, (string) $value);
    }

    /**
     * @inheritDoc
     */
    public function writeIntegerValue(?string $key, ?int $value): void
    {
        $this->writeStringValue($key, (string) $value);
    }

    /**
     * @inheritDoc
     */
    public function writeDateTimeValue(?string $key, ?DateTime $value): void
    {
        if ($value) {
            $this->writeStringValue($key, $value->format(\DateTimeInterface::RFC3339));
        }
    }

    /**
     * @inheritDoc
     */
    public function writeCollectionOfObjectValues(?string $key, ?array $values): void
    {
        throw new \RuntimeException(TextParseNode::NO_STRUCTURED_DATA_ERR_MSG);
    }

    /**
     * @inheritDoc
     */
    public function writeObjectValue(?string $key, ?Parsable $value): void
    {
        throw new \RuntimeException(TextParseNode::NO_STRUCTURED_DATA_ERR_MSG);
    }

    /**
     * @inheritDoc
     */
    public function getSerializedContent(): StreamInterface
    {
        $this->stream->rewind();
        return $this->stream;
    }

    /**
     * @inheritDoc
     */
    public function writeEnumSetValue(?string $key, ?array $values): void
    {
        throw new \RuntimeException(TextParseNode::NO_STRUCTURED_DATA_ERR_MSG);
    }

    /**
     * @inheritDoc
     */
    public function writeEnumValue(?string $key, ?Enum $value): void
    {
        if ($value) {
            $this->writeStringValue($key, $value->value());
        }
    }

    /**
     * @inheritDoc
     */
    public function writeNullValue(?string $key): void
    {
        $this->writeStringValue($key, 'null');
    }

    /**
     * @inheritDoc
     */
    public function writeAdditionalData(?array $value): void
    {
        throw new \RuntimeException(TextParseNode::NO_STRUCTURED_DATA_ERR_MSG);
    }

    /**
     * @inheritDoc
     */
    public function writeDateValue(?string $key, ?Date $value): void
    {
        $this->writeStringValue($key, (string) $value);
    }

    /**
     * @inheritDoc
     */
    public function writeTimeValue(?string $key, ?Time $value): void
    {
        $this->writeStringValue($key, (string) $value);
    }

    /**
     * @inheritDoc
     */
    public function writeDateIntervalValue(?string $key, ?DateInterval $value): void
    {
        if ($value) {
            $valueString = "P{$value->y}Y{$value->y}M{$value->d}DT{$value->h}H{$value->i}M{$value->s}S";
            $this->writeStringValue($key, $valueString);
        }
    }

    /**
     * @inheritDoc
     */
    public function writeCollectionOfPrimitiveValues(?string $key, ?array $value): void
    {
        throw new \RuntimeException(TextParseNode::NO_STRUCTURED_DATA_ERR_MSG);
    }

    /**
     * @inheritDoc
     */
    public function writeAnyValue(?string $key, $value): void
    {
        throw new \RuntimeException(TextParseNode::NO_STRUCTURED_DATA_ERR_MSG);
    }

    /**
     * @inheritDoc
     */
    public function writeBinaryContent(?string $key, ?StreamInterface $value): void
    {
        if ($value) {
            $this->writeStringValue($key, $value->getContents());
        }
    }

    /**
     * @inheritDoc
     */
    public function setOnBeforeObjectSerialization(?callable $value): void
    {
        $this->onBeforeObjectSerialization = $value;
    }

    /**
     * @inheritDoc
     */
    public function getOnBeforeObjectSerialization(): ?callable
    {
        return $this->onBeforeObjectSerialization;
    }

    /**
     * @inheritDoc
     */
    public function setOnAfterObjectSerialization(?callable $value): void
    {
        $this->onAfterObjectSerialization = $value;
    }

    /**
     * @inheritDoc
     */
    public function getOnAfterObjectSerialization(): ?callable
    {
        return $this->onAfterObjectSerialization;
    }

    /**
     * @inheritDoc
     */
    public function setOnStartObjectSerialization(?callable $value): void
    {
        $this->onStartObjectSerialization = $value;
    }

    /**
     * @inheritDoc
     */
    public function getOnStartObjectSerialization(): ?callable
    {
        return $this->onStartObjectSerialization;
    }

    public function writeByteValue(?string $key, ?Byte $value): void
    {
        $this->writeStringValue($key, (string) $value);
    }
}
